File: Minimal/tests/Brontide.Minimal.Interchange.Tests/Portable/PortablePlanObservationAndParityTests.fs
namespace Brontide.Minimal.Interchange.Tests.Portable

open System.IO
open System.Reflection
open NUnit.Framework
open Brontide.Minimal.Binding.Portable

/// C2, C7, C9, and C10: plan facts, observation completeness, realization parity, and the
/// independence the neutral layer requires.
[<TestFixture>]
type PortablePlanObservationAndParityTests() =

    let establishedPlan () =
        expectOk (
            PortableNegotiation.negotiate
                CoolingFixture.contract
                CoolingFixture.contract
                Realization.FixedDirectCall
                "minimal-direct"
                "minimal-provider"
                "fixed"
        )

    [<Test>]
    member _.``PB-53 every declared plan fact answers without invoking the provider``() =
        let plan = establishedPlan ()
        use document = readNeutral [ "schemas"; "binding-plan.json" ]

        let declared =
            document.RootElement.GetProperty("facts").EnumerateArray()
            |> Seq.collect (fun group -> group.GetProperty("entries").EnumerateArray())
            |> Seq.map (fun entry -> entry.GetProperty("id").GetString() |> str)
            |> List.ofSeq

        Assert.That(List.length declared, Is.GreaterThan 30)

        for name in declared do
            Assert.That(BindingPlan.tryFact name plan, Is.Not.Null, $"The Binding Plan answers no fact named '{name}'.")

            match BindingPlan.tryFact name plan with
            | Some _ -> ()
            | None -> Assert.Fail $"The Binding Plan answers no fact named '{name}'."

        assertAll (fun () ->
            Assert.That(BindingPlan.tryFact "constraintPolicy" plan, Is.EqualTo(Some "fail-closed"))
            Assert.That(BindingPlan.tryFact "implicitCopyPermitted" plan, Is.EqualTo(Some "false"))
            Assert.That(BindingPlan.tryFact "exceptionTransportPermitted" plan, Is.EqualTo(Some "false"))
            Assert.That(BindingPlan.tryFact "chainConjunctionTier" plan, Is.EqualTo(Some "carried"))
            Assert.That(BindingPlan.tryFact "orderingGuarantee" plan, Is.EqualTo(Some "none"))
            Assert.That(BindingPlan.tryFact "maxConcurrentRequests" plan, Is.EqualTo(Some "1")))

    [<Test>]
    member _.``PB-54 compact identifiers are plan data and never identity``() =
        let plan = establishedPlan ()
        let assignments = BindingPlan.compactAssignments plan

        Assert.That(assignments, Is.Not.Empty)

        // Every assignment names a canonical reference; the compact value only labels it inside this
        // binding.
        for assignment in assignments do
            Assert.That(assignment.Reference, Does.Contain "@")
            Assert.That(CompactId.value assignment.Compact, Is.GreaterThanOrEqualTo 0)

        let operations = expectOk (Ok(BindingPlan.tryFact "operations" plan))

        Assert.That(operations, Is.EqualTo(Some(PortableOperationRef.text CoolingFixture.setEnabled)))

        // The observation reports canonical references too.
        let host = directCoolingHost ()

        let result =
            invoke host CoolingFixture.setEnabled CoolingFixture.commandV1 (CoolingFixture.authorizedCommand "primary" true)

        shouldEqual [ CoolingFixture.setEnabled ] result.Observation.NegotiatedOperations

    [<Test>]
    member _.``PB-55 a successful interaction reports every normative field``() =
        let host = directCoolingHost ()

        let result =
            invoke host CoolingFixture.setEnabled CoolingFixture.commandV1 (CoolingFixture.authorizedCommand "primary" true)

        let observation = result.Observation

        assertAll (fun () ->
            Assert.That(Observation.completenessFailures observation, Is.Empty)
            Assert.That(observation.TerminalStatus, Is.EqualTo TerminalStatus.Succeeded)
            Assert.That(observation.FailureDomain, Is.EqualTo None)
            Assert.That(observation.SelectionReason, Is.Not.Empty)
            Assert.That(observation.NegotiatedContractVersion, Is.EqualTo 1)
            Assert.That(observation.RetryCount, Is.EqualTo 0L)
            Assert.That(observation.Interrupted, Is.False)
            Assert.That(observation.ProviderEffectCount, Is.EqualTo 1L)
            // Timing is recorded but drives no semantic decision.
            Assert.That(observation.Timing.RequestElapsedMilliseconds, Is.GreaterThanOrEqualTo 0L))

    [<Test>]
    member _.``PB-58 both realizations report the same category-level profile on success``() =
        let direct = directCoolingHost ()
        let processHost, seam = processCoolingHost ()
        use _ = seam

        let command = CoolingFixture.authorizedCommand "primary" true
        let directResult = invoke direct CoolingFixture.setEnabled CoolingFixture.commandV1 command
        let processResult = invoke processHost CoolingFixture.setEnabled CoolingFixture.commandV1 command

        shouldEqual (InteractionResult.parityProfile directResult) (InteractionResult.parityProfile processResult)

        // The realization is exactly what may differ.
        assertAll (fun () ->
            Assert.That(direct.Realization, Is.EqualTo Realization.FixedDirectCall)
            Assert.That(processHost.Realization, Is.EqualTo Realization.NegotiatedProcess)

            Assert.That(
                BindingPlan.tryFact "framing" direct.Plan,
                Is.EqualTo(Some PortableRepresentations.DirectCall)
            )

            Assert.That(
                BindingPlan.tryFact "framing" processHost.Plan,
                Is.EqualTo(Some PortableRepresentations.LengthDelimited)
            ))

    [<Test>]
    member _.``PB-59 both realizations report the same no-frame denial``() =
        let direct = directCoolingHost ()
        let processHost, seam = processCoolingHost ()
        use _ = seam

        let command = CoolingFixture.authorizedCommand "primary" true
        let denied = PortableConstraint.Atom PortableTruth.Unsatisfied

        let directResult =
            direct.Invoke(CoolingFixture.setEnabled, CoolingFixture.commandV1, command, denied).Result

        let processResult =
            processHost
                .Invoke(CoolingFixture.setEnabled, CoolingFixture.commandV1, command, denied)
                .Result

        assertAll (fun () ->
            Assert.That(directResult.FrameDecision, Is.EqualTo FrameDecision.None)
            Assert.That(processResult.FrameDecision, Is.EqualTo FrameDecision.None)
            Assert.That(directResult.ResultClass, Is.EqualTo ResultClass.Denial)
            Assert.That(processResult.ResultClass, Is.EqualTo ResultClass.Denial)

            shouldEqual (InteractionResult.parityProfile directResult) (InteractionResult.parityProfile processResult))

    [<Test>]
    member _.``PB-60 copy accounting differs by realization without breaking parity``() =
        let direct = directCoolingHost ()
        let processHost, seam = processCoolingHost ()
        use _ = seam

        let command = CoolingFixture.authorizedCommand "primary" true

        let directResult =
            direct
                .Invoke(CoolingFixture.setEnabled, CoolingFixture.commandV1, command, permitted, [ blob "cooling-profile" ])
                .Result

        let processResult =
            processHost
                .Invoke(CoolingFixture.setEnabled, CoolingFixture.commandV1, command, permitted, [ blob "cooling-profile" ])
                .Result

        assertAll (fun () ->
            Assert.That(directResult.Observation.CopyCount, Is.EqualTo 0L)
            Assert.That(processResult.Observation.CopyCount, Is.EqualTo 1L)

            // Parity still holds, because copy accounting is an excluded field with exactly this
            // stated reason, and the compared resource facts are equal.
            shouldEqual (InteractionResult.parityProfile directResult) (InteractionResult.parityProfile processResult))

    [<Test>]
    member _.``a refusal leaves both realizations in the same lifecycle state``() =
        let direct = directCoolingHost ()
        let processHost, seam = processCoolingHost ()
        use _ = seam

        // The Operation declares the host-context Fragment as required, so omitting it is refused
        // wherever the refusal is decided.
        let unattributed = CoolingFixture.command "primary" true None None None

        let directResult = invoke direct CoolingFixture.setEnabled CoolingFixture.commandV1 unattributed
        let processResult = invoke processHost CoolingFixture.setEnabled CoolingFixture.commandV1 unattributed

        assertAll (fun () ->
            Assert.That(directResult.Category, Is.EqualTo(Some ProtocolCategory.InvalidPayload))
            Assert.That(processResult.Category, Is.EqualTo(Some ProtocolCategory.InvalidPayload))
            Assert.That(directResult.Observation.ProviderEffectCount, Is.EqualTo 0L)
            Assert.That(processResult.Observation.ProviderEffectCount, Is.EqualTo 0L)
            // The declared transition to 'failed' is applied by the endpoint rather than by the
            // transport, so the same refusal leaves both realizations in the same state.
            Assert.That(processHost.State, Is.EqualTo direct.State)
            Assert.That(direct.State, Is.EqualTo LifecycleState.Failed))

    [<Test>]
    member _.``PB-62 the binding imports no other stack and the neutral layer stays data-only``() =
        let assembly = typeof<PortableBindingHost>.Assembly

        let referenced =
            assembly.GetReferencedAssemblies()
            |> Array.map (fun reference -> reference.Name)
            |> Array.append [| assembly.GetName().Name |]

        for name in referenced do
            Assert.That(name, Does.Not.StartWith "Brontide.Reference")

        // The neutral layer contains no generated stack source and no runtime helper.
        let files = Directory.GetFiles(neutralRoot, "*", SearchOption.AllDirectories)
        Assert.That(files, Is.Not.Empty)

        for file in files do
            let extension = Path.GetExtension(file: string)
            Assert.That([ ".json"; ".md" ], Contains.Item extension, $"'{file}' is not data.")

    [<Test>]
    member _.``a plan for the same contract differs only in the realization facts``() =
        let direct = establishedPlan ()

        let negotiated =
            expectOk (
                PortableNegotiation.negotiate
                    CoolingFixture.contract
                    CoolingFixture.contract
                    Realization.NegotiatedProcess
                    "minimal-direct"
                    "minimal-provider"
                    "fixed"
            )

        let differing =
            BindingPlan.factNames direct
            |> List.filter (fun name -> BindingPlan.tryFact name direct <> BindingPlan.tryFact name negotiated)
            |> List.filter (fun name -> name <> "planId")
            |> Set.ofList

        shouldEqual (Set.ofList [ "framing"; "realization"; "crossedBoundaries" ]) differing

File: Reference/tests/Brontide.Reference.Interchange.Tests/Portable/PortablePlanObservationAndParityTests.cs
using System.Collections.Immutable;
using Brontide.Reference.Experimental.Binding.Portable;

namespace Brontide.Reference.Interchange.Tests.Portable;

/// <summary>Neutral vectors PB-53 through PB-62: plan facts, observations, and realization parity.</summary>
public sealed class PortablePlanObservationAndParityTests
{
    // PB-53-PLAN-IS-FROZEN-AND-INSPECTABLE
    [Test]
    public async Task Every_declared_plan_fact_answers_by_name_without_invoking_the_provider()
    {
        var handler = new CoolingPortableHandler(CoolingPortableFixture.CreateNativeRegistry());
        await using var host = await PortableTestHarness.DirectHostAsync(handler: handler);
        var plan = host.Plan;

        var expected = new[]
        {
            "planId", "contractVersion", "component", "provider", "operations", "shapes", "fragments",
            "satisfiedRequirements", "compactIdentifierAssignments", "hostEndpoint", "providerEndpoint",
            "selectedProvider", "selectionReason", "authorityPresentationMode", "trustBoundaryCrossed",
            "noCapabilityTransfer", "constraintPolicy", "authorityDecisionPoint", "chainConjunctionTier",
            "representation", "framing", "resourceFlavors", "acceptedResourceHandles", "resourceOwnership",
            "implicitCopyPermitted", "integrityAlgorithm", "maxConcurrentRequests", "orderingGuarantee",
            "realization", "crossedBoundaries", "limits", "replayProtectionDeclared", "replayWindow",
            "lifecycleFeatures", "terminalStates", "failureDomainsObservable", "exceptionTransportPermitted"
        };

        Assert.Multiple(() =>
        {
            Assert.That(plan.FactNames, Is.EquivalentTo(expected));
            foreach (var name in expected)
            {
                Assert.That(plan.Fact(name), Is.Not.Null, $"Plan fact '{name}' did not answer.");
            }

            Assert.That(plan.Fact("constraintPolicy"), Is.EqualTo("fail-closed"));
            Assert.That(plan.Fact("implicitCopyPermitted"), Is.EqualTo("false"));
            Assert.That(plan.Fact("exceptionTransportPermitted"), Is.EqualTo("false"));
            Assert.That(plan.Fact("chainConjunctionTier"), Is.EqualTo("carried"));
            Assert.That(plan.Fact("orderingGuarantee"), Is.EqualTo("none"));
            Assert.That(plan.Fact("terminalStates"), Is.EqualTo("terminated, failed"));
            Assert.That(handler.ProviderEffectCount, Is.Zero, "Inspection invokes no provider.");
        });
    }

    // PB-54-COMPACT-IDENTIFIERS-ARE-NOT-IDENTITY
    [Test]
    public async Task The_plan_and_the_observation_report_canonical_references_never_compact_ones()
    {
        await using var host = await PortableTestHarness.DirectHostAsync();
        var plan = host.Plan;

        Assert.Multiple(() =>
        {
            Assert.That(plan.CompactIdentifierAssignments, Is.Not.Empty);
            Assert.That(plan.Fact("operations"), Does.Contain("interchange.tests.cooling.set-enabled@1"));
            Assert.That(plan.Fact("operations"), Does.Not.Match(@"^\d+$"));
            foreach (var assignment in plan.CompactIdentifierAssignments)
            {
                Assert.That(assignment.CompactId.Value, Is.InRange(0, PortableCompactIdentifier.MaxValue));
                Assert.That(assignment.Reference, Does.Contain("@"));
            }
        });
    }

    // PB-55-OBSERVATION-COMPLETE-ON-SUCCESS
    [Test]
    public async Task A_successful_interaction_reports_every_normative_field()
    {
        await using var host = await PortableTestHarness.DirectHostAsync();
        var result = await host.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true),
            PortableTestHarness.Permitted());
        var observation = result.Observation;

        Assert.Multiple(() =>
        {
            Assert.That(observation.TerminalStatus, Is.EqualTo(PortableTerminalStatus.Succeeded));
            Assert.That(observation.FailureDomain, Is.Null, "A success reports no failure domain.");
            Assert.That(observation.SelectedProvider, Is.EqualTo(CoolingPortableFixture.Provider));
            Assert.That(observation.SelectionReason, Is.Not.Empty);
            Assert.That(observation.NegotiatedOperations, Does.Contain(CoolingPortableFixture.SetEnabled));
            Assert.That(observation.NegotiatedContractVersion, Is.EqualTo(1));
            Assert.That(observation.Representation, Is.EqualTo(PortableRepresentations.PortableCborCore));
            Assert.That(observation.RetryCount, Is.Zero);
            Assert.That(observation.Interrupted, Is.False);
            Assert.That(observation.ProviderEffectCount, Is.EqualTo(1));
            Assert.That(observation.Timing.RequestElapsedMilliseconds, Is.GreaterThanOrEqualTo(0));
            Assert.That(
                observation.CorrelationMapping.HostNativeExecution.Value,
                Is.Not.EqualTo(observation.CorrelationMapping.ChannelId.Value));
        });
    }

    // PB8 B3 — an endpoint cannot fabricate a zero effect count when the peer's effect is unknowable.
    [Test]
    public async Task An_unobservable_provider_effect_uses_the_declared_unknown_form()
    {
        await using var host = await PortableTestHarness.DirectHostAsync();
        var succeeded = await host.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true),
            PortableTestHarness.Permitted());

        var unknown = succeeded.Observation with
        {
            TerminalStatus = PortableTerminalStatus.ProcessFailure,
            FailureDomain = PortableFailureDomain.Transport,
            ProviderEffectCount = null
        };

        Assert.Multiple(() =>
        {
            Assert.DoesNotThrow(unknown.RequireComplete);
            Assert.That(unknown.ParityProfile()["providerEffectCount"], Is.EqualTo("unknown"));
            Assert.Throws<InvalidOperationException>(() =>
                (unknown with { TerminalStatus = PortableTerminalStatus.Succeeded, FailureDomain = null }).RequireComplete());
        });
    }

    // PB-58-DIRECT-AND-PROCESS-PARITY-ON-SUCCESS
    [Test]
    public async Task Both_realizations_agree_on_every_field_of_the_parity_profile_for_a_success()
    {
        var direct = await RunAsync(direct: true, resources: []);
        var process = await RunAsync(direct: false, resources: []);

        Assert.Multiple(() =>
        {
            Assert.That(process.Profile, Is.EqualTo(direct.Profile));
            Assert.That(direct.Result.ResultClass, Is.EqualTo(PortableResultClass.OutcomeSucceeded));
            Assert.That(direct.Plan.Fact("realization"), Is.EqualTo("fixed-direct-call"));
            Assert.That(process.Plan.Fact("realization"), Is.EqualTo("negotiated-process"));
            Assert.That(direct.Plan.Fact("framing"), Is.EqualTo("direct-call"));
            Assert.That(process.Plan.Fact("framing"), Is.EqualTo("length-delimited"));
        });
    }

    [Test]
    public async Task The_two_plans_differ_only_in_realization_framing_and_crossed_boundaries()
    {
        var direct = await RunAsync(direct: true, resources: []);
        var process = await RunAsync(direct: false, resources: []);
        var permitted = new[] { "realization", "framing", "crossedBoundaries", "planId", "hostEndpoint" };

        var differing = direct.Plan.Facts
            .Where(fact => process.Plan.Facts[fact.Key] != fact.Value)
            .Select(fact => fact.Key)
            .ToImmutableArray();
        Assert.That(differing, Is.SubsetOf(permitted));
    }

    // PB-59-DIRECT-AND-PROCESS-PARITY-ON-DENIAL
    [Test]
    public async Task Both_realizations_make_the_same_no_frame_decision_for_a_local_denial()
    {
        var denial = PortableConstraint.AllOf(
            PortableConstraint.Atom(PortableTruth.Satisfied),
            PortableConstraint.Atom(PortableTruth.Unknown));

        await using var directHost = await PortableTestHarness.DirectHostAsync();
        var directResult = await directHost.InvokeAsync(
            CoolingPortableFixture.SetEnabled,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true),
            denial);

        var (processHost, seam) = await PortableTestHarness.ProcessHostAsync();
        await using (seam)
        await using (processHost)
        {
            var processResult = await processHost.InvokeAsync(
                CoolingPortableFixture.SetEnabled,
                CoolingPortableFixture.CommandV1,
                CoolingPortableFixture.Command("primary", enabled: true),
                denial);

            Assert.Multiple(() =>
            {
                Assert.That(processResult.ParityProfile(), Is.EqualTo(directResult.ParityProfile()));
                Assert.That(directResult.FrameDecision, Is.EqualTo(PortableFrameDecision.None));
                Assert.That(processResult.FrameDecision, Is.EqualTo(PortableFrameDecision.None));
                Assert.That(processResult.Observation.ProviderEffectCount, Is.Zero);
            });
        }
    }

    // PB-60-COPY-ACCOUNTING-DIFFERS-WITHOUT-BREAKING-PARITY
    [Test]
    public async Task Copy_accounting_differs_by_realization_while_parity_still_holds()
    {
        var direct = await RunAsync(direct: true, resources: [PortableTestHarness.Blob()]);
        var process = await RunAsync(direct: false, resources: [PortableTestHarness.Blob()]);

        Assert.Multiple(() =>
        {
            Assert.That(direct.Result.Observation.CopyCount, Is.Zero);
            Assert.That(process.Result.Observation.CopyCount, Is.EqualTo(1));
            Assert.That(process.Profile, Is.EqualTo(direct.Profile), "Copy accounting is an excluded field.");
            Assert.That(
                process.Result.Observation.ReferencedResources[0].Flavor,
                Is.EqualTo(direct.Result.Observation.ReferencedResources[0].Flavor));
            Assert.That(
                process.Result.Observation.ReferencedResources[0].IntegrityVerified,
                Is.EqualTo(direct.Result.Observation.ReferencedResources[0].IntegrityVerified));
        });
    }

    [Test]
    public void A_refused_request_leaves_the_endpoint_failed_in_the_direct_realization_too()
    {
        // The lifecycle's declared transition on a protocol error is to the failed state. Applying
        // it only in the transport would leave the two realizations in different endpoint states
        // after the same refusal.
        var endpoint = PortableTestHarness.CoolingEndpoint(PortableRealization.FixedDirectCall);
        endpoint.Establish(CoolingPortableFixture.Contract, "host");
        endpoint.SignalReady();

        Assert.Throws<PortableFaultException>(() => endpoint.Request(
            PortableChannelRequestId.New(),
            PortableOperationReference.Parse("interchange.tests.cooling.set-disabled", 1),
            null,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true),
            []));

        Assert.Multiple(() =>
        {
            Assert.That(endpoint.State, Is.EqualTo(PortableLifecycleState.Failed));
            Assert.That(endpoint.State.IsTerminal(), Is.True);
        });
    }

    // PB-62-NO-SHARED-SEMANTIC-RUNTIME
    [Test]
    public void The_portable_layer_imports_no_other_stack_runtime_and_the_neutral_layer_stays_data_only()
    {
        var referenced = typeof(PortableBindingHost).Assembly
            .GetReferencedAssemblies()
            .Select(assembly => assembly.Name ?? string.Empty)
            .ToImmutableArray();

        Assert.Multiple(() =>
        {
            Assert.That(
                referenced.Where(name => name.StartsWith("Brontide.Minimal", StringComparison.Ordinal)),
                Is.Empty);
            Assert.That(referenced.Where(name => name.StartsWith("FSharp", StringComparison.Ordinal)), Is.Empty);

            var neutral = Directory.GetFiles(PortableTestHarness.NeutralRoot, "*", SearchOption.AllDirectories);
            Assert.That(neutral, Is.Not.Empty);
            Assert.That(
                neutral.Where(path => Path.GetExtension(path) is not (".json" or ".md")),
                Is.Empty,
                "The neutral layer carries data and documentation only.");
        });
    }

    private static async Task<(PortableInteractionResult Result, PortableBindingPlan Plan, ImmutableSortedDictionary<string, string> Profile)> RunAsync(
        bool direct,
        IReadOnlyList<PortableResource> resources)
    {
        if (direct)
        {
            await using var host = await PortableTestHarness.DirectHostAsync();
            var result = await host.InvokeAsync(
                CoolingPortableFixture.SetEnabled,
                CoolingPortableFixture.CommandV1,
                CoolingPortableFixture.Command("primary", enabled: true),
                PortableTestHarness.Permitted(),
                resources);
            return (result, host.Plan, result.ParityProfile());
        }

        var (processHost, seam) = await PortableTestHarness.ProcessHostAsync();
        await using (seam)
        await using (processHost)
        {
            var result = await processHost.InvokeAsync(
                CoolingPortableFixture.SetEnabled,
                CoolingPortableFixture.CommandV1,
                CoolingPortableFixture.Command("primary", enabled: true),
                PortableTestHarness.Permitted(),
                resources);
            return (result, processHost.Plan, result.ParityProfile());
        }
    }
}
